tests/test_templates_broadcast.py

```diff
@@ -0,0 +1,243 @@
+                   ("single", Rot, [[0.1, 0.2, 0.3], [0.3, 0.2, 0.1], [0.3, 0.2, -0.1]]),
+                   ("single", T, [[], [], []]),
+                   ("double", CRX, [[0.1]]),
+                   ("double", CRot, [[0.1, 0.2, 0.3]]),
+                   ("double", CNOT, [[]]),
+                   ("double_odd", CRX, [[0.1]]),
+                   ("double_odd", CRot, [[0.3, 0.2, 0.1]]),
+                   ("double_odd", CNOT, [[]]),
+                   ]
+
+
+class TestConstructorBroadcast:
+    """Tests the broadcast template constructor."""
+
+    @pytest.mark.parametrize("structure, unitary, parameters", GATE_PARAMETERS)
+    def test_correct_queue_for_gate_unitary(self, structure, unitary, parameters):
+        """Tests that correct gate queue is created when 'block' is a single gate."""
+
+        with qml.utils.OperationRecorder() as rec:
+            broadcast(block=unitary, structure=structure, wires=range(3), parameters=parameters)
+
+        for gate in rec.queue:
+            assert isinstance(gate, unitary)
+
+    @pytest.mark.parametrize("structure, unitary, gates, parameters",
+                             [("single", ParametrizedTemplate, [RX, RY], [[0.1, 1], [0.2, 1], [0.1, 1]]),
+                              ("single", ConstantTemplate, [T, S], [[], [], []]),
+                              ("double", ParametrizedTemplateDouble, [CRX, RY], [[0.1, 1]]),
+                              ("double", ConstantTemplateDouble, [T, CNOT], [[]]),
+                              ])
+    def test_correct_queue_for_template_block(self, structure, unitary, gates, parameters):
+        """Tests that correct gate queue is created when 'block' is a template."""
+
+        with qml.utils.OperationRecorder() as rec:
+            broadcast(block=unitary, structure=structure, wires=range(3), parameters=parameters)
+
+        for idx, gate in enumerate(rec.queue):
+            i = idx % 2
+            assert isinstance(gate, gates[i])
+
+    @pytest.mark.parametrize("structure, template, kwarg, target_queue, parameters",
+                             [("single", KwargTemplate, True, [T, RY, T, RY], [[1], [2]]),
+                              ("single", KwargTemplate, False, [RY, RY], [[1], [2]]),
+                              ("double", KwargTemplateDouble, True, [T, CRX], [[1]]),
+                              ("double", KwargTemplateDouble, False, [CRX], [[1]])])
+    def test_correct_queue_for_template_block_with_keyword(self, structure, template, kwarg, target_queue, parameters):
+        """Tests that correct gate queue is created when 'block' is a template that uses a keyword."""
+
+        with qml.utils.OperationRecorder() as rec:
+            broadcast(block=template, structure=structure, wires=range(2),
+                      parameters=parameters, kwargs={'a': kwarg})
+
+        for gate, target_gate in zip(rec.queue, target_queue):
+            assert isinstance(gate, target_gate)
```

Comment:
Here checking that the same parameters were also passed might be useful

Reply:
Oh or is this the next test actually? I'm slightly confused here :sweat_smile:

Reply:
Yes, there is a separate test for parameters :)